core/src/sumeragi/network_topology.rs

```diff
@@ -5,341 +5,174 @@
+                    } else {
+                        vec![]
+                    }
+                }
+                (Role::ValidatingPeer, _) => self.sorted_peers
+                    [1..(self.min_votes_for_commit() - 1)]
+                    .iter()
+                    .map(|peer_id| peer_id.public_key.clone())
+                    .collect(),
+                (Role::ProxyTail, _) => vec![self.sorted_peers[self.min_votes_for_commit() - 1]
+                    .public_key
+                    .clone()],
+                (Role::ObservingPeer, _) => self.sorted_peers[self.min_votes_for_commit()..]
+                    .iter()
+                    .map(|peer_id| peer_id.public_key.clone())
+                    .collect(),
+            };
+            public_keys.extend(role_public_keys);
+        }
         signatures
             .into_iter()
             .filter(|signature| public_keys.contains(signature.public_key()))
             .cloned()
             .collect()
     }
-
-    /// Sorted peers that this topology has.
-    pub fn sorted_peers(&self) -> &[PeerId] {
-        &self.sorted_peers
+    /// What role does this peer have in the topology. If it is not in the toplogy
+    /// the function will return `Role::ObservingPeer`.
+    // This lint is a bad suggestion.
+    #[allow(clippy::option_if_let_else)]
+    pub fn role(&self, peer_id: &PeerId) -> Role {
+        match self.sorted_peers.iter().position(|p| p == peer_id) {
+            Some(index) if index == 0 => Role::Leader,
+            Some(index) if index < self.min_votes_for_commit() => Role::ValidatingPeer,
+            Some(index) if index == self.min_votes_for_commit() => Role::ProxyTail,
+            Some(_) => Role::ObservingPeer,
+            None => {
+                trace!(%peer_id, "Peer is not in topology.");
+                Role::ObservingPeer
+            }
+        }
     }
-
-    /// Maximum number of faulty peers that the network will tolerate.
-    #[allow(clippy::integer_division)]
-    pub fn max_faults(&self) -> usize {
-        (self.sorted_peers.len() - 1) / 3
+    /// Get leader's peer id.
+    pub fn leader(&self) -> &PeerId {
+        &self.sorted_peers[0]
     }
-
-    /// Create network topology from the list of [`WorldStateView`] peers at the given block
-    pub fn recreate(&mut self, wsv: &WorldStateView, block: HashOf<VersionedCommittedBlock>) {
-        let wsv_peers = wsv.peers_ids().iter().map(|id| id.clone());
-
-        *self = Topology::builder()
-            .with_peers(wsv_peers)
-            .at_block(block)
-            .build(0)
-            .expect("Invalid topology");
+    /// Get proxy tail's peer id.
+    pub fn proxy_tail(&self) -> &PeerId {
+        &self.sorted_peers[self.min_votes_for_commit()]
+    }
+    /// Add or remove peers from the topology.
+    pub fn update_peer_list(&mut self, new_peer_list: &[PeerId]) {
+        // The complexity here is to not two peer_ids that are different but have the same key.
```

Comment:
something's wrong with the comment

Reply:
also, isn't this impossible? Shouldn't public key be a unique identifier for a peer? @appetrosyan @Erigara

Reply:
yes, @SamHSmith please check comment on `PartialEq` implementation for `PeerId`

Reply:
Yea I have now rewritten it.